### ISSUE TITLE
Expose electron version to extension code

### ISSUE THREAD

**Issue by salterok** [Microsoft/vscode-languageserver-node#434]
In general, extension that uses binary dependency may want to read electron version to properly prepare such dependencies to work in provided environment.

I'm want to integrate telemetry into extension to better understand users behavior.
Google's `stackdriver` chosen for this purpose, but there is issue that stop me from deploy extension.
Stackdriver libs using another one called `grpc` which is binary module and it requires to be built upon specific elector version (as long as extension code use electrons `require`).

So to sum up, when my extension downloaded/updated by vscode, it starts to download precompiled  `grpc` binary, but there i need to get electron version vscode compiled with.

**Comment by dbaeumer** [comment]
Is `process.versions` not working for you. Should give you something like:

![capture](https://user-images.githubusercontent.com/1931590/47644373-3d87de00-db6e-11e8-95f1-0beee77ac5a0.png)

**Comment by dbaeumer** [comment]
Please ping if this is not what you were looking for.

**Comment by salterok** [comment]
@dbaeumer Thanks for response. 

This exactly what i need but `process.versions` give me next result:
![image](https://user-images.githubusercontent.com/1340183/47665859-ccb2e700-dbaa-11e8-8070-012a3ffa6aec.png)


I see it differs a lot from your example so will try to provide more info (maybe i'm missing some essential part).

Code i'm working on, running as "Language Client" (i.e. using VS Code API directly).
Process has environment variable `ELECTRON_RUN_AS_NODE` set to "1".
VS Code info:
![image](https://user-images.githubusercontent.com/1340183/47665739-780f6c00-dbaa-11e8-9475-9a156ff2e6e9.png)

**Comment by dbaeumer** [comment]
@salterok I see your point. I will reopen the issue.

I discussed this with @bpasero our Electron expert and he has some ideas how to solve this. He will follow up.

**Comment by bpasero** [comment]
@salterok you should be looking at the `modules` entry to decide which version to download. This is after all the one that drives the decision if the native module can be loaded or not. I just verified that I can use a native module that I compiled with node `8.12.0` (which uses `V8 62`) in our Electron environment that uses `V8 61` with node `8.9.3`.

**Comment by dbaeumer** [comment]
@bpasero thanks !!

**Comment by salterok** [comment]
@bpasero thanks, I followed you recommendation and even made it work, but solution may be error-prone as I can tell.
So let me describe my concerns:

That binary module `grpc` use `node-pre-gyp` to populate itself, which require *runtime* and *target* args (among others) to properly download/build module.
So, to get version of `grpc` that works with specified abi version I used `node-abi` module. It gives me electron version *1.8.0* for abi *57*. 
While it seems to be binary compatible with electron 2.0 (which is also abi v57) I doubt about possible issues in such version mismatch.

Even if using module, built on top of electron@1.8, is ok under electron@2.0 there is also issue with how `node-pre-gyp` prepare binaries and where they stored. 
To solve this I need to do one of the following:
 * copy/move binary to the right place after `node-pre-gyp` complete execution but before require'ing `grpc`.
 * set `process.versions.electron` manually (to version which `node-abi` gives) before running `node-pre-gyp`.

After all that's said you may close this issue as i got partial solution. 
But in my opinion if extension is run under electron it should know that from environment, to minimize incompatibility issues, so exposing `process.versions.electron` would be ideal solution.

Thanks you @dbaeumer and @bpasero for help!

**Comment by bpasero** [comment]
Yeah, if `node-pre-gyp` is involved then I think that solution I proposed is not ideal. If we add the electron version to `process` it would only be available in the extension, but not in the language server (unless we would drag it over there too), would that be an issue?

**Comment by salterok** [comment]
If I understand docs properly language server is intended to be separate process (non restricted to NodeJs) so extension author should be able to configure such process in a proper way.
On the other hand extension is started by vscode in which case author has no control over environment.
So i think it should be ok that only extension gets proper versions in `process` as long as developer can control language server creation.

**Comment by dbaeumer** [comment]
The extension author can control the start of the server even if it is a node server. Instead of describing it via a literal the developer can pass a function that is called to start the server to give extension developers full control over this.

**Comment by dbaeumer** [comment]
@salterok any objection to close the issue here. Setting the electron version for the extension host is something that needs to be done in VS Code itself and we would need an issue here: https://github.com/Microsoft/vscode/issues

**Comment by salterok** [comment]
No. I'll open another one in main `vscode` repo.

Thanks again for help!
